### PR TITLE
PoseStamped vector specialization

### PR DESCRIPTION
**Pull request by tonynajjar**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | https://github.com/ros-navigation/navigation2/issues/4388 |
| Primary OS tested on | (Ubuntu, MacOS, Windows) |
| Robotic platform tested on | (Steve's Robot, gazebo simulation of Tally, hardware turtlebot) |
| Does this PR contain AI generated software? | (No; Yes and it is marked inline in the code) |

---

## Description of documentation updates required from your changes

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by tonynajjar** [comment]
PR is not done, will ping once ready for review. I just need to open a PR to remember finishing it :)

**Comment by SteveMacenski** [comment]
@tonynajjar hidy ho, any chance we can get this in soon? :upside_down_face:

**Comment by tonynajjar** [comment]
I thrive on pings, I'll try to resume it this week. I got stuck on something but forgot on what.

**Comment by tonynajjar** [comment]
Pretty much ready for a first review, one thing I couldn't get to the bottom yet of is why test_goal_updated_controller suddenly fails with these changes. (At least locally, let's see in CI)

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_utils.hpp [hunk 1]
```diff
@@ -102,6 +104,70 @@ inline geometry_msgs::msg::PoseStamped convertFromString(const StringView key)
...
+ * @brief Parse XML string to nav_msgs::msg::Path
+ * @param key XML string
+ * @return nav_msgs::msg::Path
+ */
+template<>
+inline nav_msgs::msg::Path convertFromString(const StringView key)
+{
+  // 9 real numbers separated by semicolons
+  auto parts = BT::splitString(key, ';');
+  if ((parts.size() - 9) % 7 == 0) {
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_utils.hpp:148]: Can you explain this? I see 3 in header (header, sec, nanosec) for `parts.size() - 3` & the message sizes of 10 (2 parts for timestamp: sec and nanosec) Or maybe 9 if the stamp is converted into a single float?

**Review by SteveMacenski — COMMENTED** [review]
I retriggered CI, if it fails again, then its a real issue -- those tests are pretty simple API use so if that's broken, there's a real possibility the actual BT node is broken too and not just a test update

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_utils.hpp (hunk 1 above)
Comment (tonynajjar) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_utils.hpp]: > Can you explain this?

I can't, I don't know what I was thinking, I fixed it now. There should be 2 (path header) + 9 (PoseStamped fields) x N

File: nav2_behavior_tree/plugins/decorator/goal_updated_controller.cpp [hunk 2]
```diff
@@ -37,18 +36,18 @@ BT::NodeStatus GoalUpdatedController::tick()
     // Reset since we're starting a new iteration of
     // the goal updated controller (moving from IDLE to RUNNING)
 
-    BT::getInputOrBlackboard("goals", goals_);
-    BT::getInputOrBlackboard("goal", goal_);
+    goals_ = config().blackboard->get<std::vector<geometry_msgs::msg::PoseStamped>>("goals");
```
Comment (tonynajjar) [inline nav2_behavior_tree/plugins/decorator/goal_updated_controller.cpp:39]: I don't know why, but `BT::getInputOrBlackboard("goals", goals_);` is not getting the goals set in the test whereas this version does.

File: nav2_behavior_tree/plugins/decorator/goal_updated_controller.cpp [hunk 3]
```diff
@@ -37,18 +36,18 @@ BT::NodeStatus GoalUpdatedController::tick()
...
+    goal_ = config().blackboard->get<geometry_msgs::msg::PoseStamped>("goal");
```
Comment (tonynajjar) [inline nav2_behavior_tree/plugins/decorator/goal_updated_controller.cpp:40]: weirdly enough, here it works with `BT::getInputOrBlackboard("goal", goal_);` but I changed it to keep it consistent

File: nav2_behavior_tree/test/plugins/decorator/test_goal_updated_controller.cpp [hunk 4]
```diff
@@ -36,8 +36,6 @@ class GoalUpdatedControllerTestFixture : public nav2_behavior_tree::BehaviorTree
     poses1.push_back(goal1);
     config_->blackboard->set("goal", goal1);
     config_->blackboard->set("goals", poses1);
-    config_->input_ports["goals"] = "";
```
Comment (tonynajjar) [inline nav2_behavior_tree/test/plugins/decorator/test_goal_updated_controller.cpp:39]: @SteveMacenski these were causing the test to fail, I see they were added [there](https://github.com/ros-navigation/navigation2/pull/4298/files#diff-b9f432eb1b03868819ba84ee7c21c4b98a8cc0d4a589587c46a0e0d1120d7f05R39-R40) by you I believe, do you remember what were you solving with these lines?

Comment (SteveMacenski) [reply nav2_behavior_tree/test/plugins/decorator/test_goal_updated_controller.cpp:39]: There were issues in the 24.04 migration about ports not being set, so I set them all to something in the tests. It doesn't need to be `""`, but needs to be initialized as something to stop BT.CPP update from throwing an uninitialized port error due to how the tests are structured

**Review by SteveMacenski — APPROVED** [review]
Is this good to go?

**Comment by tonynajjar** [comment]
CI is happy I'm happy

**Pull request merged**